src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java

```diff
@@ -408,24 +427,56 @@ FireStoreDirectoryEntry computeDirectory(
         String fullPath = fireStoreUtils.getFullPath(dirEntry.getPath(), dirEntry.getName());
         List<FireStoreDirectoryEntry> enumDir = directoryDao.enumerateDirectory(firestore, snapshotId, fullPath);
 
-        // Recurse to compute results from underlying directories
         List<FireStoreDirectoryEntry> enumComputed = new ArrayList<>();
-        for (FireStoreDirectoryEntry dirItem : enumDir) {
-            if (dirItem.getIsFileRef()) {
-                // Read the file metadata to get the size and checksum. We do a bit of a hack and copy
-                // the size and checksums into the in-memory dirItem. That way we only compute on the directory
-                // objects.
-                FireStoreFile file =
-                    fileDao.retrieveFileMetadata(firestore, dirItem.getDatasetId(), dirItem.getFileId());
-
-                dirItem
-                    .size(file.getSize())
-                    .checksumMd5(file.getChecksumMd5())
-                    .checksumCrc32c(file.getChecksumCrc32c());
-
-                enumComputed.add(dirItem);
-            } else {
-                enumComputed.add(computeDirectory(firestore, snapshotId, dirItem, updateBatch));
+
+        // Recurse to compute results from underlying directories
+        try (Stream<FireStoreDirectoryEntry> stream = enumDir.stream()) {
+            enumComputed.addAll(stream
+                .filter(f -> !f.getIsFileRef())
+                .map(f -> {
+                    try {
+                        return computeDirectory(firestore, snapshotId, f, updateBatch);
+                    } catch (InterruptedException e) {
+                        throw new ApiException("Error computing directory metadata", e);
+                    }
+                })
+                .collect(Collectors.toList())
+            );
+        }
+
+        // Collect metadata for file objects in the directory
+        try (Stream<FireStoreDirectoryEntry> stream = enumDir.stream()) {
+            // Group FireStoreDirectoryEntry objects by dataset Id to process one dataset at a time
+            final Map<String, List<FireStoreDirectoryEntry>> fileRefsByDatasetId = stream
+                .filter(FireStoreDirectoryEntry::getIsFileRef)
+                .collect(Collectors.groupingBy(FireStoreDirectoryEntry::getDatasetId));
+
+            for (Map.Entry<String, List<FireStoreDirectoryEntry>> entry : fileRefsByDatasetId.entrySet()) {
+                // Retrieve the file metadata from Firestore
+                final List<FireStoreFile> fireStoreFiles =
+                    fileDao.batchRetrieveFileMetadata(firestore, entry.getKey(), entry.getValue());
+
+                // Index by fileId to be able to look up
```

Comment:
Why can't you just put the `dirItem` into the `enumComputed` here instead of building a separate structure just to re-look it up?

Reply:
That's kind of the crux of the change here.  We were able do that before because we were looking up the metadata 1 by 1 (which was the slow bit).  This goal here was to batch but the problem is that it adds complexity because you need to resolve the results afterwards.  So I get the bulk results from the metadata look up but then I need to take the original `FireStoreDirectoryEntry` objects and look up what was returned from firebase to augment them with the appropriate size and metadata.

Reply:
Ah interesting.  That's a really good point.  I generally get nervous about positional dependency (since nothing stops somebody from changing the underlying implementation to something non-order-maintaining) but that makes sense with the current way the code is and saves us having to build up this separate map.   I'll maybe add a comment to `batchRetrieveFileMetadata` to make it clear that order in = order out

Reply:
The documentation on that could certainly use a fix. We depend on that order in = order out property in most of the places where we use it. And it is coded to maintain that property.

Reply:
Totally.  That's really good to know.  I'll update